userspace/falco/app_actions/init_falco_engine.cpp

```diff
@@ -57,25 +56,80 @@ void application::configure_output_format()
 	}
 }
 
-application::run_result application::init_falco_engine()
+void application::add_source_to_engine(const std::string& src)
 {
-	configure_output_format();
+	auto src_info = m_state->sources.at(src);
+	std::shared_ptr<gen_event_filter_factory> filter_factory = nullptr;
+	std::shared_ptr<gen_event_formatter_factory> formatter_factory = nullptr;
 
-	// Create "factories" that can create filters/formatters for syscalls
+	if (src == falco_common::syscall_source)
+	{
+		filter_factory = std::shared_ptr<gen_event_filter_factory>(
+			new sinsp_filter_factory(src_info->inspector.get()));
+		formatter_factory = std::shared_ptr<gen_event_formatter_factory>(
+			new sinsp_evt_formatter_factory(src_info->inspector.get()));
+	}
+	else
+	{
+		auto &filterchecks = m_state->sources.at(src)->filterchecks;
+		filter_factory = std::shared_ptr<gen_event_filter_factory>(
+			new sinsp_filter_factory(src_info->inspector.get(), filterchecks));
+		formatter_factory = std::shared_ptr<gen_event_formatter_factory>(
+			new sinsp_evt_formatter_factory(src_info->inspector.get(), filterchecks));
+	}
+
+	if(m_state->config->m_json_output)
+	{
+		formatter_factory->set_output_format(gen_event_formatter::OF_JSON);
+	}
 
-	// libs requires raw pointer, we should modify libs to use reference/shared_ptr
-	std::shared_ptr<gen_event_filter_factory> syscall_filter_factory(new sinsp_filter_factory(m_state->inspector.get()));
+	src_info->engine_idx = m_state->engine->add_source(
+		src, filter_factory, formatter_factory);
+}
 
-	// libs requires raw pointer, we should modify libs to use reference/shared_ptr
-	std::shared_ptr<gen_event_formatter_factory> syscall_formatter_factory(new sinsp_evt_formatter_factory(m_state->inspector.get()));
+application::run_result application::init_falco_engine()
+{
+	// add all non-syscall event sources in engine
+	for (const auto& src : m_state->loaded_sources)
+	{
+		if (src != falco_common::syscall_source)
+		{
+			// we skip the syscall as we want it to be the one added for last
+			// in the engine. This makes the source index assignment easier.
+			add_source_to_engine(src);
+		}
+	}
 
-	m_state->syscall_source_idx = m_state->engine->add_source(falco_common::syscall_source, syscall_filter_factory, syscall_formatter_factory);
-	
-	if(m_state->config->m_json_output)
+	// add syscall as last source
+	add_source_to_engine(falco_common::syscall_source);
```

Comment:
Do we want to always add `syscall_source` unconditionally, without checking if it is loaded?

Reply:
Good point I'm gonna try to respond to this point and some of the other questions below. So, for event sources we maintain in the app state three things:
- `loaded_sources`: a set of names all loaded event sources. By default this contains `syscall` that is provided out-of-the-box by Falco, and the event sources coming from the loaded plugins.
- `enabled_sources`: a set of names all event sources and that are enabled for event processing (either in capture/offline or in live mode). By default this is the same as `loaded_sources`, and can then be changed with the `--enable-source` and `--disable-source` options.
- `sources`: a list indexed by source name that contains all information mapped to a given source. For example, we map 1 inspector for each source name, 1 filtercheck list, and so on

The action of adding an event source to the rule engine relies on the `loaded_sources` list, because the rule engine must be aware of all sources no matter if they are actually enabled for event processing or not, because otherwise it will not be able to load the configured rulesets. Doing otherwise will deteriorate the UX as Falco would stop with error at startup if configured with even a single rule attached to a non-enabled event source. This is also why the `syscall` event source is added unconditionally.

Reply:
Note that the "all enabled by default" and the `--disable-source` options behavior are the same as when we had k8s_audit support bundled into Falco. I thought it made sense to reproduce the same UX.

Reply:
Makes sense! Thanks Jason!